compiler/optimizer/VPHandlers.cpp

```diff
@@ -5802,22 +5804,43 @@ TR::Node *constrainAcall(OMR::ValuePropagation *vp, TR::Node *node)
                               && !vp->_arrayCloneCalls.find(vp->_curTree))
                         {
                         vp->_arrayCloneCalls.add(vp->_curTree);
-                        vp->_arrayCloneTypes.add(constraint->getClass());
+                        vp->_arrayCloneTypes.add(new (vp->trStackMemory()) OMR::ValuePropagation::ArrayCloneInfo(constraint->getClass(), true));
                         }
                      }
                   }
+               // Dynamic object clone is enabled only with FLAGS_IN_CLASS_SLOT build flag enabled
+               // as currenty codegen anewarray evaluator only supports this case for object header initialization.
+               // Even though all existing supported build config has this falg set, this ifdef serves as a safety precaution.
+#ifdef J9VM_INTERP_FLAGS_IN_CLASS_SLOT
                else if ( constraint->getClassType()
                          && constraint->getClassType()->asResolvedClass() )
                   {
                   newTypeConstraint = TR::VPResolvedClass::create(vp, constraint->getClass());
+                  if (vp->trace())
+                     traceMsg(vp->comp(), "Object Clone: Resolved Class of node %p \n", node);
                   if (enableDynamicObjectClone
                       && constraint->getClassType()->isArray() == TR_no
                       && !vp->_objectCloneCalls.find(vp->_curTree))
                      {
+                     if (vp->trace())
+                        traceMsg(vp->comp(), "Object Clone: Resolved Class of node %p object clone\n", node);
                      vp->_objectCloneCalls.add(vp->_curTree);
                      vp->_objectCloneTypes.add(new (vp->trStackMemory()) OMR::ValuePropagation::ObjCloneInfo(constraint->getClass(), false));
                      }
+                  // Currently enabled for X86 as the required codegen support is implemented on X86 only.
+                  // Remove the condition as other platforms receive support.
+                  else if (TR::Compiler->target.cpu.isX86()
```

Comment:
Is it better to have a "capability" based query in the codegen object here instead of checking that we are on X86 explicitly ? When the codegens other than X86 implement the necessary function they can return true to the capability based query and this code won't have to change.